lib/srv/uacc/uacc.h

```diff
@@ -0,0 +1,136 @@
+/*
+Copyright 2021 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#ifndef UACC_C
+#define UACC_C
+
+#include <unistd.h>
+#include <stdint.h>
+#include <string.h>
+#include <utmp.h>
+#include <errno.h>
+#include <stdbool.h>
+
+#define SELECT_UTMP_PATH(x) x != NULL ? x : _PATH_UTMP
+#define SELECT_WTMP_PATH(x) x != NULL ? x : _PATH_WTMP
+
+int UACC_UTMP_MISSING_PERMISSIONS = 1;
+int UACC_UTMP_WRITE_ERROR = 2;
+int UACC_UTMP_READ_ERROR = 3;
+int UACC_UTMP_FAILED_OPEN = 4;
+int UACC_UTMP_ENTRY_DOES_NOT_EXIST = 5;
+int UACC_UTMP_FAILED_TO_SELECT_FILE = 6;
+
+// I initially attempted to use the login/logout BSD functions but ran into a string of unexpected behaviours such as
+// errno being set to undocument values along with wierd return values in certain cases. They also modify the utmp database
+// in a way we don't want. We want to insert a USER_PROCESS entry directly before we do PAM/cgroup setup and launch the shell
+// without any middleman. I brought this information with my issues and requirements to the IRC and was told that I would be
+// better off using the setutent/pututline low level manipulation functions - joel.
+
+// The max byte length of the C string representing the TTY name.
+static int max_len_tty_name() {
+    return UT_LINESIZE;
+}
+
+// Low level C function to add a new USER_PROCESS entry to the database.
+// This function does not perform any argument validation.
+static int uacc_add_utmp_entry(const char *utmp_path, const char *wtmp_path, const char *username, const char *hostname, const int32_t remote_addr_v6[4], const char *tty_name, const char *id, int32_t tv_sec, int32_t tv_usec) {
+    if (utmpname(SELECT_UTMP_PATH(utmp_path)) != 0) {
+        return UACC_UTMP_FAILED_TO_SELECT_FILE;
+    }
+    struct utmp entry;
+    entry.ut_type = USER_PROCESS;
+    strncpy((char*) &entry.ut_line, tty_name, UT_LINESIZE);
+    strncpy((char*) &entry.ut_id, id, sizeof(entry.ut_id));
+    entry.ut_pid = getpid();
+    strncpy((char*) &entry.ut_host, hostname, sizeof(entry.ut_host));
+    strncpy((char*) &entry.ut_user, username, sizeof(entry.ut_user));
```

Comment:
I think the size is incorrect here. You want to copy one less than the maximum size (and have it null terminated). In particular for the username argument, because it could be attacker controlled.

https://us-cert.cisa.gov/bsi/articles/knowledge/coding-practices/strncpy-and-strncat